envoy/envoy.go

```diff
@@ -1,81 +1,378 @@
+	if err != nil {
+		return err
+	}
+
+	serverTemplate, err := newCertificateTemplate(serverKey, opts.hosts, notBefore, notAfter, false)
+	if err != nil {
+		return err
+	}
+
+	_, serverCertBytes, err := makeCertificate(serverTemplate, caCert, publicKey(serverKey), caKey)
+	if err != nil {
+		return err
+	}
+
+	// save ca certificate
+	err = savePEM(certsDir, "cacert.pem", []*pem.Block{{Type: "CERTIFICATE", Bytes: caCertBytes}}, defaultFileFlags, 0600)
+	if err != nil {
+		return err
+	}
+
+	caKeyBytes, err := x509.MarshalPKCS8PrivateKey(caKey)
+	if err != nil {
+		return fmt.Errorf("unable to marshal ca private key: %v", err)
+	}
+	err = savePEM(certsDir, "cakey.pem", []*pem.Block{{Type: "PRIVATE KEY", Bytes: caKeyBytes}}, defaultFileFlags, 0600)
+	if err != nil {
+		return err
+	}
+
+	// save server certificate
+	err = savePEM(certsDir, "cert.pem", []*pem.Block{{Type: "CERTIFICATE", Bytes: serverCertBytes}}, defaultFileFlags, 0600)
+	if err != nil {
+		return err
+	}
+
+	serverKeyByes, err := x509.MarshalPKCS8PrivateKey(serverKey)
+	if err != nil {
+		return fmt.Errorf("unable to marshal server private key: %v", err)
+	}
+
+	err = savePEM(certsDir, "key.pem", []*pem.Block{{Type: "PRIVATE KEY", Bytes: serverKeyByes}}, defaultFileFlags, 0600)
+	if err != nil {
+		return err
+	}
+
+	// save fullchain
+	err = savePEM(certsDir, "fullchain.pem", []*pem.Block{
+		{Type: "CERTIFICATE", Bytes: serverCertBytes},
+		{Type: "CERTIFICATE", Bytes: caCertBytes},
+	}, defaultFileFlags, 0600)
+	if err != nil {
+		return err
 	}
-	l.Info("starting Envoy container...")
-	// TODO(meling) use docker api to run image: "docker run --name=envoy -p 8080:8080 --net=host ag_envoy"
-	out, err := exec.Command("/bin/sh", "./envoy/envoy.sh").Output()
+
+	log.Printf("certificates successfully generated at: %s", certsDir)
+	return nil
+}
+
+// loadConfigEnv loads the  envoy config from the environment variables.
+// It will not override a variable that already exists.
+// Consider the .env file to set development vars or defaults.
+func loadConfigEnv(withTLS bool, config *CertificateConfig) (*EnvoyConfig, error) {
+	err := godotenv.Load(env)
 	if err != nil {
-		l.Fatal("failed to execute bash script", zap.Error(err))
+		return nil, err
 	}
-	l.Debug("envoy.sh", zap.String("output", string(out)))
+
+	return newEnvoyConfig(os.Getenv("DOMAIN"), os.Getenv("SERVER_HOST"), os.Getenv("GRPC_PORT"), os.Getenv("HTTP_PORT"), withTLS, config)
 }
 
-// hasEnvoyImage returns true if the docker client has the latest Envoy image.
-func hasEnvoyImage(ctx context.Context, l *zap.Logger, cli *client.Client) bool {
-	l.Debug("no running Envoy container found")
-	images, err := cli.ImageList(ctx, types.ImageListOptions{})
+// TODO: improve parameter handling when generating certificates (keyType, etc).
+// TODO: save certs at: /etc/ssl/certs and private keys at: /etc/ssl/private by default.
+func main() {
+	flag.BoolVar(&genConfig, "genconfig", false, "generate envoy config")
```

Comment:
We should consider if this should be a standalone tool (as is now), or if this should be moved to another command line tool inside `cmd/` folder. Right now, we have too many standalone tools.

Another option is to make this part of the `quickfeed` command itself, but running it something like this:

```sh
% quickfeed gen-envoy-config --tls --cert cert.pem --key private.pem
```

Which would also entail that:

```sh
% quickfeed run -service.url $DOMAIN
```

At any rate, even a standalone tool should avoid the `--genconfig` flag, and instead have it be a command like `genconfig`.

But this comment is mainly added for context and brainstorming... Maybe we should create a separate issue for this.

Reply:
I moved it to `cmd` and change the flags. But I let it as standalone tool for now, but we can change this in another PR later.